pygeos/_geometry.pyx

```diff
@@ -321,3 +323,120 @@ def collections_1d(object geometries, object indices, int geometry_type = 7):
             geom_idx_1 += collection_size[coll_idx]
 
     return result
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def polygons_1d(object shells, object holes, object indices):
+    cdef Py_ssize_t hole_idx_1 = 0
+    cdef Py_ssize_t poly_idx = 0
+    cdef unsigned int n_holes = 0
+    cdef Py_ssize_t poly_hole_idx = 0
+    cdef GEOSGeometry *shell = NULL
+    cdef GEOSGeometry *hole = NULL
+    cdef GEOSGeometry *poly = NULL
+
+    # Cast input arrays and define memoryviews for later usage
+    shells = np.asarray(shells, dtype=object)
+    if shells.ndim != 1:
+        raise TypeError("shells is not a one-dimensional array.")
+
+    # Cast input arrays and define memoryviews for later usage
+    holes = np.asarray(holes, dtype=object)
+    if holes.ndim != 1:
+        raise TypeError("holes is not a one-dimensional array.")
+
+    indices = np.asarray(indices, dtype=np.int32)
+    if indices.ndim != 1:
+        raise TypeError("indices is not a one-dimensional array.")
+
+    if holes.shape[0] != indices.shape[0]:
+        raise ValueError("holes and indices do not have equal size.")
+
+    if shells.shape[0] == 0:
+        # return immediately if there are no geometries to return
+        return np.empty(shape=(0, ), dtype=object)
+
+    if (indices >= shells.shape[0]).any():
+        raise ValueError("Some indices are of bounds of the shells array.")  
+
+    if np.any(indices[1:] < indices[:indices.shape[0] - 1]):
+        raise ValueError("The indices should be sorted.")  
+
+    cdef Py_ssize_t n_poly = shells.shape[0]
+    cdef object[:] shells_view = shells
+    cdef object[:] holes_view = holes
+    cdef int[:] indices_view = indices
+
+    # get the holes count per polygon (this raises on negative indices)
+    cdef int[:] hole_count = np.bincount(indices, minlength=n_poly).astype(np.int32)
+
+    # A temporary array for the holes that will be given to CreatePolygon
+    # Its size equals max(hole_count) to accomodate the largest polygon.
+    temp_holes = np.empty(shape=(np.max(hole_count), ), dtype=np.intp)
+    cdef np.intp_t[:] temp_holes_view = temp_holes
+
+    # The final target array
+    result = np.empty(shape=(n_poly, ), dtype=object)
+    cdef object[:] result_view = result
+
+    with get_geos_handle() as geos_handle:
+        for poly_idx in range(n_poly):
+            n_holes = 0
+
+            # get the shell
+            if PyGEOS_GetGEOSGeometry(<PyObject *>shells_view[poly_idx], &shell) == 0:
+                raise TypeError(
+                    "One of the arguments is of incorrect type. Please provide only Geometry objects."
+                )
+
+            # return None for missing shells (ignore possibly present holes)
+            if shell == NULL:
+                result_view[poly_idx] = PyGEOS_CreateGeometry(NULL, geos_handle)
+                continue
```

Comment:
```suggestion
                hole_idx_1 += hole_count[poly_idx]
                continue
```

See failing test below

Reply:
Good catch! I corrected it and added the test.